[Payments NOX in LYS] Improve WooPayments plugin state detection

vladolaru opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

We make consistent use of the backend-provided WooPayments state data both in the LYS sidebar and content sections. The backend-provided data already has the mechanisms to detect plugins in test or development directories ([here](https://github.com/woocommerce/woocommerce/blob/94c260bba3989d14883a5278d0b4afd3d9790efa/plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders.php#L1392-L1392)).

We also:
- always surface step errors if they are provided by the backend (this means that they are unhandled or retries failed)
- improved the standardization of step error data
- improved i18n to use placeholders for the WooPayments brand to better guard our brand against wrongful translations.
- teach AI agents (Claude) about handling i18n in JS/TS

Closes WOOPLUG-5903

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Spin up a new JurassicNinja store with Jetpack Beta, WC on this PR's branch, WC Beta Tester (here's a [direct create link](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-beta-tester&branches.woocommerce=update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection))
2. Go to the site's dashboard, click on WooCommerce, skip the profiler, and set the store location to the US.
3. Go to WooCommerce > Home and click on the "Launch your store" task.
4. In LYS, click on the "Set up payments" task.
5. Both the sidebar step and the main content CTA button should ask you to "install" WooPayments:
<img width="1426" height="750" alt="Screenshot 2025-11-24 at 17 44 20" src="https://github.com/user-attachments/assets/bb1cb8ad-bb9f-47aa-bb66-a30befce5683" />
6. Click on "Install" and, after WooPayments is installed, you will be taken to the payment methods step, with the "Install WooPayments" step crossed out:
<img width="1428" height="751" alt="Screenshot 2025-11-24 at 17 45 44" src="https://github.com/user-attachments/assets/6524e337-c271-4276-8bae-7b0228b700bd" />

7. Close LYS, go to your WP admin dashboard, Plugins > Installed plugins, and deactivate&delete the WooPayments plugin.
8. Grab a WooPayments zip (e.g., from [here](https://github.com/Automattic/woocommerce-payments/releases/tag/10.3.0-test-1)), extract it locally, change the directory name to `woocommerce-payments-dev`, and zip it again.
9. Go to Plugins > Add plugin and install the zip you crafted, but don't activate WooPayments.
10. Go back to WooCommerce > Home, click on the "Launch your store" task, then click on "Set up payments"
11. Both the sidebar step and the button CTA should say "Enable", not "Install":
<img width="1425" height="751" alt="Screenshot 2025-11-24 at 17 54 36" src="https://github.com/user-attachments/assets/865c1037-f01f-4cfd-8deb-a34d69fc3e18" />

12. Hard-refresh the page, and after initially using "Install", the copy should change to "Enable" (once the data is fetched).
13. Click on "Enable," and WooPayments should be activated, and the payment methods setup step should be shown.
14. Go to Plugins > Installed plugins and make sure that the test WooPayments is the plugin that was activated (no other version is installed).

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [ ] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

vladolaru committed:
Ensure main context is using a single source of truth for plugin state and is reactive

vladolaru committed:
Ensure sidebar is using a single source of truth for plugin state and is reactive

vladolaru committed:
Add changelog

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
refac: Use a helper for fetching the Payments task

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
Use the provider slug when installing WooPayments

vladolaru committed:
Fix missing location when removing shell gateways

vladolaru committed:
Don't log when receiving default cuntry value

vladolaru committed:
Standardize WooPayments step errors

vladolaru committed:
Revert to using useSelect on the pluginStore

vladolaru committed:
Surface backend-provided step errors

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders/WooPayments/WooPaymentsService.php:
```diff
@@ -2031,6 +2031,19 @@ private function standardize_onboarding_step_details( array $step_details, strin
+				);
+			}
+		}
+		$step_details['errors'] = $standardized_errors;
```
No problem here. This API is only used internally.

vladolaru committed:
More defensive step error standardization

vladolaru committed:
Guard against unmounted components

vladolaru committed:
Avoid rapid re-renders

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
Better standardization of step errors

vladolaru committed:
test: Add unit tests for step error standardization changes

vladolaru committed:
refac: Use placeholders for brand names in i18n strings

vladolaru committed:
ai: Teach Claude to handle i18n strings in JS/TS

vladolaru committed:
Fix markdown linting

vladolaru commented on .ai/skills/woocommerce-dev-cycle/js-i18n-patterns.md:
```diff
@@ -0,0 +1,298 @@
+# JavaScript/TypeScript i18n Patterns
```
Just teaching AI...

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/sidebar/components/payments-mobile-header.tsx:
```diff
@@ -64,7 +64,10 @@ export const PaymentsMobileHeader = ( props: SidebarComponentProps ) => {
 			/>
 			<h1 className="mobile-header__title">
-				{ __( 'Set up WooPayments', 'woocommerce' ) }
+				{
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/sidebar/components/payments-sidebar.tsx:
```diff
@@ -95,7 +117,10 @@ export const PaymentsSidebar = ( props: SidebarComponentProps ) => {
 			} }
 		>
-			{ __( 'Set up WooPayments', 'woocommerce' ) }
+			{
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/payments-welcome/strings.tsx:
```diff
@@ -18,9 +18,13 @@ export default {
-		__(
-			'By using WooPayments you agree to our <a1>Terms of Service</a2> and acknowledge that you have read our <a2>Privacy Policy</a2>. Discount will be applied to payments processed via WooPayments upon completion of installation, setup, and connection. ',
-			'woocommerce'
+		sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/payments-welcome/strings.tsx:
```diff
@@ -42,9 +46,13 @@ export default {
-		__(
-			'By using WooPayments you agree to our <a1>Terms of Service</a2> (including WooPay <a3>merchant terms</a3>) and acknowledge that you have read our <a2>Privacy Policy</a2>. Discount will be applied to payments processed via WooPayments upon completion of installation, setup, and connection. ',
-			'woocommerce'
+		sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/payments-welcome/strings.tsx:
```diff
@@ -86,17 +94,29 @@ export default {
-	paymentOptions: __(
-		'WooPayments is pre-integrated with all popular payment options',
-		'woocommerce'
+	paymentOptions: sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/payments-welcome/strings.tsx:
```diff
@@ -86,17 +94,29 @@ export default {
-		intro: __(
-			'Note that the extension hasn’t been installed. This will simply dismiss our limited time offer. Please take a moment to tell us why you’d like to dismiss the WooPayments offer.',
-			'woocommerce'
+		title: sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/payments-welcome/strings.tsx:
```diff
@@ -110,9 +130,10 @@ export default {
-		moreInfoLabel: __(
-			'I need more information about WooPayments',
-			'woocommerce'
+		moreInfoLabel: sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/components/modals/woo-payments-update-required-modal.tsx:
```diff
@@ -80,7 +80,11 @@ export const WooPaymentsUpdateRequiredModal = ( {
 							onClick={ handleUpdateWooPayments }
 						>
-							{ __( 'Update WooPayments', 'woocommerce' ) }
+							{ sprintf(
```
Just making sure our brand is guarded during i18n.

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/business-verification/index.tsx:
```diff
@@ -70,6 +71,15 @@ export const BusinessVerificationStep: React.FC = () => {
 		<div className="settings-payments-onboarding-modal__step-business-verification">
 			<WooPaymentsStepHeader onClose={ closeModal } />
 			<div className="settings-payments-onboarding-modal__step-business-verification-content">
+				{ currentStep?.errors && currentStep.errors.length > 0 && (
```
If there are unhandled errors sent by the backend, always show them. It is better to inform the user and ourselves).

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/finish/index.tsx:
```diff
@@ -14,12 +14,22 @@ import './style.scss';
 		<>
 			<WooPaymentsStepHeader onClose={ closeModal } />
 			<div className="settings-payments-onboarding-modal__step--content">
+				{ currentStep?.errors && currentStep.errors.length > 0 && (
```
If there are unhandled errors sent by the backend, always show them. It is better to inform the user and ourselves).

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/payment-methods-selection/index.tsx:
```diff
@@ -232,6 +232,15 @@ export default function PaymentMethodsSelection() {
 						) }
 					</div>
 				</div>
+				{ currentStep?.errors && currentStep.errors.length > 0 && (
```
If there are unhandled errors sent by the backend, always show them. It is better to inform the user and ourselves).

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/test-or-live-account/index.tsx:
```diff
@@ -60,6 +60,16 @@ const TestOrLiveAccountStep = () => {
 								</p>
 							</div>
 						</div>
+						{ currentStep?.errors &&
```
If there are unhandled errors sent by the backend, always show them. It is better to inform the user and ourselves).

vladolaru commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/wpcom-connection/index.tsx:
```diff
@@ -24,6 +24,15 @@ export const JetpackStep: React.FC = () => {
 	return (
 		<>
 			<WooPaymentsStepHeader onClose={ closeModal } />
+			{ currentStep?.errors && currentStep.errors.length > 0 && (
```
If there are unhandled errors sent by the backend, always show them. It is better to inform the user and ourselves).

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders/WooPayments/WooPaymentsService.php:
```diff
@@ -738,12 +738,24 @@ private function sanitize_onboarding_step_error( array $error ): array {
 			'context' => array(),
 		);
 
+		// Move all extra keys (not code, message, context) into the context.
```
Better handling of step errors.

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders/PaymentGateway.php:
```diff
@@ -275,8 +275,10 @@ public function get_supports_list( WC_Payment_Gateway $payment_gateway ): array
 		$country_code = strtoupper( sanitize_text_field( $country_code ) );
-		// Validate the country code format - expect ISO 3166-1 alpha-2.
-		if ( strlen( $country_code ) !== 2 || ! ctype_upper( $country_code ) ) {
+		// Validate the country code format, if provided - expect ISO 3166-1 alpha-2.
```
Removed unnecessary logging when the country code is sent as an empty string (which is valid).

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/Payments.php:
```diff
@@ -78,7 +78,7 @@ final public function init( PaymentsProviders $payment_providers, ExtensionSugge
 		$payment_gateways = $this->providers->get_payment_gateways( $for_display );
 		if ( ! $for_display && $remove_shells ) {
-			$payment_gateways = $this->providers->remove_shell_payment_gateways( $payment_gateways );
+			$payment_gateways = $this->providers->remove_shell_payment_gateways( $payment_gateways, $location );
```
The country/location should always be sent.

vladolaru committed:
More linting fixes

vladolaru committed:
Handle nested context in step error data and flatten it

vladolaru committed:
Handle edge case when a single error is provided for a step

vladolaru committed:
test: Add unit tests for step error handling of single errors

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders/WooPayments/WooPaymentsService.php:
```diff
@@ -2031,6 +2051,44 @@ private function standardize_onboarding_step_details( array $step_details, strin
 				}
 			}
 		}
+		// Standardize errors to be a list of arrays with `code`, `message`, and optional extra keys.
```
No worries. We are the only ones using these APIs.

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

mordeth reviewed (dismissed):
Thank you for working on this improvement!

✅ Code changes are looking good!
✅ Testing instructions are checking out correctly!
⚠️ I’ve noticed that the Install step disappears from the list whenever the WooPayments plugin is already installed.

https://github.com/user-attachments/assets/15935065-2e45-4263-be6d-eaa1e3cfa5dd

Pre-approving the PR.

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
fix(payments): add defensive error handling and accessibility to onboarding steps

The WooPayments onboarding error notices lacked defensive coding patterns
that could cause runtime errors when error.message was undefined, and
were missing explicit ARIA attributes for screen reader compatibility.

Add optional chaining and fallback messages to error notice rendering in
wpcom-connection and test-or-live-account steps. Add role="alert" for
better screen reader announcement compatibility with older assistive
technologies. Add try/catch error handling to PaymentsSidebar async task
fetching. Add unit tests for getPaymentsTaskFromLysTasklist function.

Refs WOOPLUG-5903

vladolaru committed:
fix(payments): add defensive checks and improve accessibility

The setup-payments-context selectors called array methods on potentially
null/undefined returns from getPaymentProviders(), getActivePlugins(), and
getInstalledPlugins(). This could cause runtime errors if the stores
returned unexpected values during loading or error states.

Add Array.isArray() guards to all three selectors and document that
useSelect handles store subscriptions internally (no empty dependency
array needed).

The business-verification step accessed error.message without guarding
against missing message property. Add optional chaining and fallback text.
Also add role="alert" for better screen reader compatibility.

Fix translator comment in PaymentsRemindMeLater to use positional format
(1:) matching the %1$s placeholder in the translatable string.

Refs WOOPLUG-5903

vladolaru committed:
test(payments): improve WooPayments error standardization test assertions

Test assertions were incomplete - they verified individual keys existed but
didn't ensure expected exceptions were actually thrown or that the error
structure was complete.

Add explicit fail() calls to verify exceptions are thrown, add assertions
for all required error keys (code, message, context), verify context is
not empty when it should contain moved keys, and remove overly strict
count assertion that was brittle.

Refs WOOPLUG-5903

vladolaru commented on plugins/woocommerce/src/Internal/Admin/Settings/PaymentsProviders/WooPayments/WooPaymentsService.php:
```diff
@@ -2031,6 +2051,44 @@ private function standardize_onboarding_step_details( array $step_details, strin
+				);
+			}
+
+			$standardized_errors[] = $this->sanitize_onboarding_step_error( $error );
```
No problem because we are the only consumer, for now.

vladolaru committed:
fix(payments): add missing useSelect dependencies array

TypeScript enforces that useSelect requires a second argument for the
dependencies array. The first useSelect call was missing this argument,
causing build failures.

Add empty dependencies array since the selector subscribes to store state
internally and re-runs automatically when isFetching or providers change.

Refs WOOPLUG-5903

vladolaru committed:
fix(payments): prevent state flipping during WooPayments slug resolution

When a test/beta WooPayments version is installed under a different slug,
the previous implementation could cause state flipping:
1. Initial load: checks if official slug is active → false
2. After fetch: checks if test slug is active → true
3. UI experiences unexpected false→true transition

Wait for provider data to load before exposing meaningful state. During
loading, render children with stable false values for isWooPaymentsActive
and isWooPaymentsInstalled. This prevents UI from making decisions based
on potentially incorrect state that would flip after load completes.

Refs WOOPLUG-5903

vladolaru committed:
fix(payments): validate plugin slug before installation

The plugin slug from the provider data was passed directly to the plugin
installer without validation. This could allow unexpected or malicious
input to reach the installation system if the API response was tampered.

Add validatePluginSlug helper that ensures:
- Value is a non-empty string (trimmed)
- Contains only valid characters (lowercase letters, numbers, dashes, underscores)
- Rejects control characters, slashes, spaces, and other special characters

If validation fails, returns undefined which triggers the safe fallback
to the official WooPayments slug constant.

Refs WOOPLUG-5903

vladolaru committed:
fix(payments): correct interpolation tag mismatch in welcome strings

The Terms of Service links used mismatched interpolation tags `<a1>...</a2>`
which could cause rendering issues with the interpolateComponents function.

Change closing tags from `</a2>` to `</a1>` to properly close the opening
`<a1>` tags in both the standard and WooPay terms strings.

Refs WOOPLUG-5903

vladolaru committed:
test(payments): add explicit fail assertions for expected exceptions

Tests expecting exceptions should call `$this->fail()` after the code
that should throw, ensuring the test fails if no exception is thrown.

Add `$this->fail('Expected ... not thrown')` calls and improve exception
verification in `onboarding_test_account_init` error handling test.

Refs WOOPLUG-5903

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -148,67 +151,88 @@ export const PaymentsContent = ( {} ) => {
-				if ( ! isWooPaymentsInstalled ) {
-					// Record the extension installation event.
-					recordPaymentsEvent( 'provider_installed', {
+	// Cleanup on unmount to prevent state updates after component is unmounted.
```
Only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/sidebar/components/payments-sidebar.tsx:
```diff
@@ -42,11 +44,43 @@ export const PaymentsSidebar = ( props: SidebarComponentProps ) => {
-	const { context } = props;
-	const payments_task = context.tasklist?.tasks?.find(
-		( task ) => task.id === 'payments'
+	// Fetch payments task using getPaymentsTaskFromLysTasklist helper
```
Only used internally

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -36,7 +36,7 @@ const InstallWooPaymentsStep = ( {
 	isPluginInstalled,
 }: {
-	installWooPayments: () => void;
+	installWooPayments: ( slug: string | undefined ) => void;
```
Only used internally.

vladolaru committed:
fix(payments): remove incorrect exception message assertion in test

The catch block in test_error_sanitization_flattens_nested_context was
asserting that the exception message contains 'test_error', but the
WP_Error code ('test_error') is different from its message ('Test error
message').

The catch block's purpose is simply to allow the test to proceed to the
actual error sanitization assertions below - it doesn't need to validate
the exception content.

Refs WOOPLUG-5903

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
fix(payments): always show Install step in LYS payments sidebar

The Install/Enable WooPayments step was only shown when the
`wooPaymentsRecentlyActivated` flag was true. This caused the step to
disappear after initial activation, making the sidebar incomplete.

Remove the conditional wrapper so the step always renders with
`isStepComplete={true}` when WooPayments is active, providing consistent
UI feedback regardless of activation timing.

Refs WOOPLUG-5903

vladolaru committed:
test(payments): add unit tests for PaymentsSidebar Install step visibility

Add comprehensive tests for the InstallWooPaymentsStep component in the
Launch Your Store payments sidebar. Tests verify:

- Install step renders with isStepComplete=false when WooPayments is not active
- Install step renders with isStepComplete=true when WooPayments is active
- Loading placeholder shows when WooPayments is active and loading
- Correct text ("Install" vs "Enable") based on installation state
- Additional onboarding steps render when WooPayments is active

Refs WOOPLUG-5903

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/sidebar/components/payments-sidebar.tsx:
```diff
@@ -42,11 +44,43 @@ export const PaymentsSidebar = ( props: SidebarComponentProps ) => {
-	const payments_task = context.tasklist?.tasks?.find(
-		( task ) => task.id === 'payments'
+	// Fetch payments task using getPaymentsTaskFromLysTasklist helper
+	const [ payments_task, setPaymentsTask ] = useState< TaskType | undefined >(
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -31,12 +31,47 @@ import {
 } from '~/settings-payments/constants';
 
+/**
+ * Validates and sanitizes a plugin slug.
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/data/setup-payments-context.tsx:
```diff
@@ -79,6 +121,26 @@ export const SetUpPaymentsProvider: React.FC< {
 
+	// Wait for slug resolution to prevent state flipping.
+	// During initial load, we don't know if a test/beta version is installed under a different slug.
+	// Rendering children without waiting could show incorrect UI that then flips after load.
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/data/setup-payments-context.tsx:
```diff
@@ -79,6 +121,26 @@ export const SetUpPaymentsProvider: React.FC< {
+		return (
+			<SetUpPaymentsContext.Provider
+				value={ {
+					isWooPaymentsActive: false,
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -148,67 +189,88 @@ export const PaymentsContent = ( {} ) => {
-					// Record the extension installation event.
-					recordPaymentsEvent( 'provider_installed', {
+	const installWooPayments = useCallback(
+		( realPluginSlug: string | undefined ) => {
```
This is only used internally.

vladolaru committed:
chore: add CLAUDE.local.md to gitignore

Local Claude configuration files should not be tracked in version control
as they contain user-specific settings.

vladolaru committed:
feat(payments): improve error display in business verification step

The previous implementation only displayed the first error message,
leaving users unaware of additional issues during onboarding setup.

Display up to 3 individual error messages for actionable feedback.
When more errors occur, show a summary count with a generic message
to avoid overwhelming the UI. Add normalizeErrorMessage() to safely
handle empty or whitespace-only messages with a fallback.

Extract OnboardingError interface to types.ts with documentation
explaining the backend contract where both fields are always present
but may be empty strings.

Refs WOOPLUG-5903

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -148,67 +189,88 @@ export const PaymentsContent = ( {} ) => {
-					// Record the extension installation event.
-					recordPaymentsEvent( 'provider_installed', {
+	const installWooPayments = useCallback(
+		( realPluginSlug: string | undefined ) => {
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/data/setup-payments-context.tsx:
```diff
@@ -79,6 +121,26 @@ export const SetUpPaymentsProvider: React.FC< {
+		return (
+			<SetUpPaymentsContext.Provider
+				value={ {
+					isWooPaymentsActive: false,
```
This is only used internally.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -31,12 +31,47 @@ import {
 	isPluginInstalled,
 }: {
-	installWooPayments: () => void;
+	installWooPayments: ( slug: string | undefined ) => void;
```
This is only used internally.

vladolaru committed:
fix(payments): correct test mocks for CI

Fix two test issues that caused CI failures:

- Fix SiteHub mock path in PaymentsSidebar test to match actual import
  (~/customize-store/site-hub instead of ~/customize-store/assembler-hub/site-hub)
- Move eslint-disable comment to correct line in BusinessVerificationStep test
  to properly suppress the unused variable warning

Refs WOOPLUG-5903

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -149,66 +152,73 @@ export const PaymentsContent = ( {} ) => {
+						reason: 'error',
 					} );
-				}
-				// Note: The provider extension activation is tracked from the backend (the `provider_extension_activated` event).
-
-				setIsPluginInstalling( false );
-			} )
-			.catch( ( response: { errors: Record< string, string > } ) => {
-				// Handle errors during installation
-				let eventName = 'provider_extension_installation_failed';
-				if ( isWooPaymentsInstalled ) {
-					eventName = 'provider_extension_activation_failed';
-				}
-				recordPaymentsEvent( eventName, {
-					provider_id: wooPaymentsProviderId,
-					suggestion_id: wooPaymentsSuggestionId,
-					provider_extension_slug: wooPaymentsExtensionSlug,
-					from: 'lys',
-					source: wooPaymentsOnboardingSessionEntryLYS,
-					reason: 'error',
+					createNoticesFromResponse( response );
+					setIsPluginInstalling( false );
 				} );
-				createNoticesFromResponse( response );
-				setIsPluginInstalling( false );
-			} );
-	}, [
-		setIsPluginInstalling,
-		installAndActivatePlugins,
-		refreshStoreData,
-		setWooPaymentsRecentlyActivated,
-	] );
+		},
+		[
+			setIsPluginInstalling,
+			installAndActivatePlugins,
+			refreshStoreData,
+			setWooPaymentsRecentlyActivated,
+		]
+	);
```
Fixed. `isWooPaymentsInstalled` is now included in the dependency array (line 268).

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/main-content/pages/payments-content.tsx:
```diff
@@ -149,66 +152,73 @@ export const PaymentsContent = ( {} ) => {
+					}
+					// Note: The provider extension activation is tracked from the backend (the `provider_extension_activated` event).
+
+					setIsPluginInstalling( false );
+				} )
+				.catch( ( response: { errors: Record< string, string > } ) => {
+					// Handle errors during installation
+					let eventName = 'provider_extension_installation_failed';
+					if ( isWooPaymentsInstalled ) {
+						eventName = 'provider_extension_activation_failed';
+					}
+					recordPaymentsEvent( eventName, {
 						provider_id: wooPaymentsProviderId,
 						suggestion_id: wooPaymentsSuggestionId,
 						provider_extension_slug: wooPaymentsExtensionSlug,
 						from: 'lys',
 						source: wooPaymentsOnboardingSessionEntryLYS,
+						reason: 'error',
 					} );
-				}
-				// Note: The provider extension activation is tracked from the backend (the `provider_extension_activated` event).
-
-				setIsPluginInstalling( false );
-			} )
-			.catch( ( response: { errors: Record< string, string > } ) => {
-				// Handle errors during installation
-				let eventName = 'provider_extension_installation_failed';
-				if ( isWooPaymentsInstalled ) {
-					eventName = 'provider_extension_activation_failed';
-				}
-				recordPaymentsEvent( eventName, {
-					provider_id: wooPaymentsProviderId,
-					suggestion_id: wooPaymentsSuggestionId,
-					provider_extension_slug: wooPaymentsExtensionSlug,
-					from: 'lys',
-					source: wooPaymentsOnboardingSessionEntryLYS,
-					reason: 'error',
+					createNoticesFromResponse( response );
+					setIsPluginInstalling( false );
 				} );
```
Fixed. Added `isMountedRef` pattern with cleanup on unmount (lines 192-199) and guards in both `.then()` (lines 223-225) and `.catch()` (lines 247-249) handlers.

vladolaru commented on plugins/woocommerce/client/admin/client/launch-your-store/data/setup-payments-context.tsx:
```diff
@@ -35,21 +37,34 @@ export const SetUpPaymentsProvider: React.FC< {
+
+	// Use the real plugin slug from the provider, falling back to the official slug.
+	const wooPaymentsPluginSlug =
+		wooPaymentsProvider?.plugin?.slug ?? wooPaymentsExtensionSlug;
```
Fixed. The `useSelect` now retrieves `isFetching` (line 46) and returns an `isSlugResolved` flag (lines 55-66). The provider waits for slug resolution before exposing real context values (lines 127-142), preventing UI flicker from slug changes during initial load.

vladolaru committed:
test(payments): improve PaymentsSidebar test reliability

- Reset mockGetPaymentsTaskFromLysTasklist to default value in beforeEach
  to prevent state leakage between tests
- Replace setTimeout-based async handling with waitFor from RTL
- Update "Enable WooPayments" test to actually verify the text content
  instead of just checking element presence

Refs WOOPLUG-5903

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

vladolaru committed:
test(payments): minor mock and assertion improvements

- Make onClick optional in Button mock to match real component interface
- Use not.toHaveClass() instead of className.not.toContain() for consistency

Refs WOOPLUG-5903

vladolaru committed:
Merge branch 'trunk' into update/WOOPLUG-5903-payments-nox-lys-woopayments-plugin-state-detection

mordeth reviewed (approved):
✅ Code changes are looking rock solid!
✅ Testing instructions are checking out correctly!

https://github.com/user-attachments/assets/1d93a1cf-f9f9-46fc-a5b7-5b01b4417b6a
